### ISSUE TITLE
Already initialized constant Rouge::Lexers:Shell::KEYWORDS

### ISSUE THREAD

**Issue by stanhu** [jneen/rouge#296]
Now that GitLab is using Rouge, it appears that Rouge is loading files automatically, causing Rails to warn with errors:

```
/Users/stanhu/.rbenv/versions/2.1.6/lib/ruby/gems/2.1.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:20: warning: already initialized constant Rouge::Lexers::Shell::KEYWORDS
/Users/stanhu/.rbenv/versions/2.1.6/lib/ruby/gems/2.1.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:20: warning: previous definition of KEYWORDS was here
/Users/stanhu/.rbenv/versions/2.1.6/lib/ruby/gems/2.1.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:25: warning: already initialized constant Rouge::Lexers::Shell::BUILTINS
/Users/stanhu/.rbenv/versions/2.1.6/lib/ruby/gems/2.1.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:25: warning: previous definition of BUILTINS was here
```

These loads happen twice:

1. https://github.com/jneen/rouge/blob/master/lib/rouge/lexers/powershell.rb#L5 
1. https://github.com/jneen/rouge/blob/master/lib/rouge.rb#L43

Removing these lines seems to make the warnings go away. Is it possible to remove these automatic loads?

**Comment by jacobvosmaer** [comment]
Sounds like a use case for Ruby's `require` instead of `load`.

**Comment by rumpelsepp** [comment]
Among others I have replaced these load calls with require relative. I forgot to file a pull request for this.
https://github.com/rumpelsepp/rugments/commit/e6bdd4eeed47670710d48f53c4b7759a35f1f348

**Comment by stanhu** [comment]
@rumpelsepp Thanks. Took your work and made it work for the specs and `rackup` application in #298.

**Comment by rumpelsepp** [comment]
:+1:

**Comment by tejasbubane** [comment]
This happens in `irb` as well. 
gem version `1.9.1`.

```
/lib/rouge/lexers/shell.rb:20: warning: already initialized constant Rouge::Lexers::Shell::KEYWORDS
/lib/rouge/lexers/shell.rb:20: warning: previous definition of KEYWORDS was here
/lib/rouge/lexers/shell.rb:25: warning: already initialized constant Rouge::Lexers::Shell::BUILTINS
/lib/rouge/lexers/shell.rb:25: warning: previous definition of BUILTINS was here
```

**Comment by nhoizey** [comment]
I also have the issue with Jekyll and Kramdown…

**Comment by JoshCheek** [comment]
Why did they have the loads in the first place? Also, this is introduced in https://github.com/jneen/rouge/commit/c8bf4a85be897327d2bcccd570fe66706dca1227

This is an alternative that maintains the `load` behaviour (again, IDK why it's there, a guess would be for reloading in rack when working tweaking the visual themes, but are better ways to achieve that, it does seem like the right solution would be to use require, except for in the case of rack, where we would want to add autoloading)

```diff
diff --git a/lib/rouge/lexers/shell.rb b/lib/rouge/lexers/shell.rb
index 8b7134e..2a7acea 100644
--- a/lib/rouge/lexers/shell.rb
+++ b/lib/rouge/lexers/shell.rb
@@ -17,12 +17,14 @@ module Rouge
         text.shebang?(/(ba|z|k)?sh/) ? 1 : 0
       end

-      KEYWORDS = %w(
+      KEYWORDS = "" unless defined? KEYWORDS
+      KEYWORDS.replace %w(
         if fi else while do done for then return function
         select continue until esac elif in
       ).join('|')

-      BUILTINS = %w(
+      BUILTINS = "" unless defined? BUILTINS
+      BUILTINS.replace %w(
         alias bg bind break builtin caller cd command compgen
         complete declare dirs disown echo enable eval exec exit
         export false fc fg getopts hash help history jobs kill let
```

**Comment by gpakosz** [comment]
Could this be fixed with a `1.9.2` maintenance release please?

**Comment by JoshCheek** [comment]
Aye, I've dealt with it by [disallowing 1.9.1](https://github.com/JoshCheek/what-we-ve-got-here-is-an-error-to-communicate/blob/3886aa779a0e6c2d3004a8603fa5daf82a63e018/what_weve_got_here_is_an_error_to_communicate.gemspec#L13)

**Comment by envygeeks** [comment]
Thanks @JoshCheek I totally did not know we could do what you did in a Gemspec.

**Comment by sc0ttwad3** [comment]
Still getting the issue... during Jekyll generations currently... (on Windows boxes as well -- below).  Has @JoshCheek found a workaround?  Steps involved? 

    C:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:20: warning: already initialized constant Rouge::Lexers::Shell::KEYWORDS
    C:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:20: warning: previous definition of KEYWORDS was here
    C:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:25: warning: already initialized constant Rouge::Lexers::Shell::BUILTINS
    C:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/rouge-1.9.1/lib/rouge/lexers/shell.rb:25: warning: previous definition of BUILTINS was here

**Comment by stanhu** [comment]
I suspect the loads were used to make the visual test `rackup` work. If you make a change in the code, the loads will pull in the new code automatically. https://github.com/jneen/rouge/issues/298 emulates this behavior by manipulating the `LOADED_FEATURES` value.

**Comment by sc0ttwad3** [comment]
Great Ill give it a try, thx.

**Comment by vincentwoo** [comment]
:+1:

**Comment by wolf99** [comment]
:+1:

**Comment by divoxx** [comment]
Why is `load` used everywhere instead of `require` ?

**Comment by gpakosz** [comment]
Which commit closes this issue please?

**Comment by jneen** [comment]
65c69f0. It's fixed as of 1.10.0.

**Comment by gpakosz** [comment]
Thank you!
